src/ty/mod.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019
+//     by  Centrality Investments Ltd.
+//     and Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::tm_std::*;
+
+use crate::{
+	form::{CompactForm, Form, MetaForm},
+	IntoCompact, MetaType, Metadata, Registry,
+};
+use derive_more::From;
+use serde::Serialize;
+
+mod composite;
+mod fields;
+mod path;
+mod variant;
+
+pub use self::{composite::*, fields::*, path::*, variant::*};
+
+/// The possible types a Rust value could have.
+///
+/// Describes the
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, From, Debug, Serialize)]
+#[serde(bound = "F::TypeId: Serialize")]
+#[serde(rename_all = "camelCase")]
+pub enum Type<F: Form = MetaForm> {
+	/// A composite type (e.g. a struct or a tuple)
+	Composite(TypeComposite<F>),
+	/// A variant type (e.g. an enum)
+	Variant(TypeVariant<F>),
+	/// A slice type with runtime known length.
+	Slice(TypeSlice<F>),
+	/// An array type with compile-time known length.
+	Array(TypeArray<F>),
+	/// A tuple type.
+	Tuple(TypeTuple<F>),
+	/// A Rust primitive type.
+	Primitive(TypePrimitive),
+}
+
+impl IntoCompact for Type {
+	type Output = Type<CompactForm>;
+
+	fn into_compact(self, registry: &mut Registry) -> Self::Output {
+		match self {
+			Type::Composite(composite) => composite.into_compact(registry).into(),
+			Type::Variant(variant) => variant.into_compact(registry).into(),
+			Type::Slice(slice) => slice.into_compact(registry).into(),
+			Type::Array(array) => array.into_compact(registry).into(),
+			Type::Tuple(tuple) => tuple.into_compact(registry).into(),
+			Type::Primitive(primitive) => primitive.into(),
+		}
+	}
+}
+
+/// A primitive Rust type.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Debug)]
+#[serde(rename_all = "lowercase")]
+pub enum TypePrimitive {
+	/// `bool` type
+	Bool,
+	/// `char` type
+	Char,
+	/// `str` type
+	Str,
+	/// `u8`
+	U8,
+	/// `u16`
+	U16,
+	/// `u32`
+	U32,
+	/// `u64`
+	U64,
+	/// `u128`
+	U128,
+	/// `i8`
+	I8,
+	/// `i16`
+	I16,
+	/// `i32`
+	I32,
+	/// `i64`
+	I64,
+	/// `i128`
+	I128,
+}
+
+/// An array type.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Debug)]
+#[serde(bound = "F::TypeId: Serialize")]
+pub struct TypeArray<F: Form = MetaForm> {
+	/// The length of the array type.
+	pub len: u16,
```

Comment:
I think we should make this a `u32` even though it is unlikely that an array with more than 2^16 fields is ever going to be used.